Introduce a <model> loading process - Part 3

eddydas opened the pull request:
#### 73946828b41b9d2d1cb13e062ed95c4ba85affa6
<pre>
Introduce a &lt;model&gt; loading process - Part 3
<a href="https://bugs.webkit.org/show_bug.cgi?id=243874">https://bugs.webkit.org/show_bug.cgi?id=243874</a>
<a href="https://rdar.apple.com/122051588">rdar://122051588</a> (Create new com.apple.WebKit.Model process)

Reviewed by Elliott Williams.

Added Xcode target for the actual ModelProcess XPC service.
Added sandbox profile (which inherits directly from WebProcess for now).
Actually trigger the launching of the process when &lt;model&gt; tag is present
when the user enables the ModelProcess path.
Stops the existing UIProcess based rendering path when the user enables
ModelProcess path.

* .gitignore:
Removed a rule that doesn&apos;t seem to apply to anything recent Xcode generates.
Having it around makes git ignore the model process sandbox profile unexpectedly.

* Source/WebKit/Configurations/ModelService.xcconfig: Added.
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelService/Info-embedded.plist: Added.
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Model.sb.in: Added.
* Source/WebKit/Scripts/compile-sandbox.sh:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp:
(WebKit::ModelProcessProxy::updateProcessAssertion):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeHostIOS.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/UIProcess/WebContextClient.cpp:
(WebKit::WebContextClient::modelProcessDidCrash): Deleted.
* Source/WebKit/UIProcess/WebContextClient.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::modelProcessExited):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::createModelProcessConnection):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.mm:

Canonical link: <a href="https://commits.webkit.org/274408@main">https://commits.webkit.org/274408@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c8efc31410bafbacc1b6bb428f1b7178d79da73b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39009 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17942 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41362 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41543 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20817 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15290 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39583 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15111 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/33820 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13125 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/38874 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13080 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42820 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35410 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35076 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/38917 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13821 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11404 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37142 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15427 "Built successfully") | | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8731 "") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15088 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/14913 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on .gitignore:
```diff
@@ -1,4 +1,3 @@
-*.mode*
```
The sadness of a case sensitive version control system on a case insensitive file system

hortont424 replied:
You have no idea how long this boggled me for

aprotyas commented on Source/WebKit/Scripts/process-entitlements.sh:
```diff
@@ -440,6 +440,11 @@ fi
 
+function ios_family_process_model_entitlements()
+{
+    ios_family_process_webcontent_shared_entitlements
```
Is this subject to change? I assumed the model loading process could have a smaller subset of entitlements.

eddydas replied:
yes! this will be revisited together with its sandbox profile once we have put in the actual code to render the model

emw-apple commented on Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelService/Info-embedded.plist:
```diff
@@ -0,0 +1,47 @@
+<plist version="1.0">
+<dict>
+	<key>CFBundleGetInfoString</key>
+	<string>${BUNDLE_VERSION}, Copyright 2003-2022 Apple Inc.</string>
```
nit: 2022

emw-apple commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -17015,6 +17066,29 @@
+				2D56C1BD28A367A00081BD25 /* PBXTargetDependency */,
+			);
+			name = Model;
+			productName = Networking;
```
This probably doesn't do anything, since you're setting PRODUCT_NAME in ModelService.xcconfig, but it's evidence of your copy-paste. (And it looks like the GPU process is the same way, lol)
```suggestion
			productName = Model;
```

eddydas replied:
Ha! Interesting... there are no UI in Xcode that shows this at all it seems..? Thanks for catching this

emw-apple commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -19210,7 +19383,9 @@
-			platformFilter = ios;
+			platformFilters = (
+				ios,
+			);
```
Just confirming that this should be safe to commit. AFAICT, the `platformFilters` key has been recognized since Xcode 13, which is as far back as our support goes.

emw-apple commented on Source/WebKit/Configurations/ModelService.xcconfig:
Are we ready to start building and installing the model process on all platforms that use XPC services? Despite it being unused by default, and the feature flag controlling it being internal-only?

Since the ENABLE(MODEL_PROCESS) platform flag is visionOS only, maybe we should do:
```
SKIP_INSTALL = YES;
SKIP_INSTALL[sdk=xr*] = NO;
```

eddydas replied:
Good call! Thanks

emw-apple reviewed (approved):
Looks great, assuming the builds you're testing internally work! LMK if I can help with any validation.

eddydas committed:
Introduce a <model> loading process - Part 3
https://bugs.webkit.org/show_bug.cgi?id=243874
rdar://122051588 (Create new com.apple.WebKit.Model process)

Reviewed by Elliott Williams.

Added Xcode target for the actual ModelProcess XPC service.
Added sandbox profile (which inherits directly from WebProcess for now).
Actually trigger the launching of the process when <model> tag is present
when the user enables the ModelProcess path.
Stops the existing UIProcess based rendering path when the user enables
ModelProcess path.

* .gitignore:
Removed a rule that doesn't seem to apply to anything recent Xcode generates.
Having it around makes git ignore the model process sandbox profile unexpectedly.

* Source/WebKit/Configurations/ModelService.xcconfig: Added.
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/ModelProcess/EntryPoint/Cocoa/XPCService/ModelService/Info-embedded.plist: Added.
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.Model.sb.in: Added.
* Source/WebKit/Scripts/compile-sandbox.sh:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp:
(WebKit::ModelProcessProxy::updateProcessAssertion):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeHostIOS.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/UIProcess/WebContextClient.cpp:
(WebKit::WebContextClient::modelProcessDidCrash): Deleted.
* Source/WebKit/UIProcess/WebContextClient.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::modelProcessExited):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::createModelProcessConnection):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.mm:

Canonical link: https://commits.webkit.org/274408@main
